### PR TITLE
feat(docs) a beginner's guide to RBP

### PR DESCRIPTION
**Pull request by KarandikarMihir**
This is for #1274 

There's a growing need for beginner-friendly docs so that they can adopt this boilerplate as quickly as possible. I've chalked out a basic diagram for better understanding of RBP. There's no point in explaining things which are too complex or things which are included prerequisites. Because the point of this document is outlining workflow in a broader sense.

I have not included unit-testing part of RBP, because I remember, as a beginner I was more interested in the actual working rather than testing. 

I'd like to re-iterate that I'm not a native English speaker. So there are going to be minor if not horrible mistakes in the docs. We usually carry out bug-bashes. Let there be a grammar-bash! :collision:

**_Please point out-_**
- [x] Grammar/style mistakes
- [x] Any other feature that you want to document
- [x] Mistakes/enhancements in the diagram
- [x] Corrections/suggestions in the content.
- [x] Change in headings/subheadings

### CODE REVIEW

**Comment by gihrig** [comment]
@KarandikarMihir Looking really good!

I have only given it a quick skim, but I see no major issues with the text. I hope to give it a more in-depth review, but if I don't get to that before you're done, it can be merged anyway. Really it's that good :smile:

Some thought:
1. I think the diagram could be improved by splitting it into a few more focussed diagrams
- How does the 'app/index.html' relate to `build/index.html`. For example, the source code in app is not executable as-is, it must be compiled by webpack and it's plugins.
- The diagram has an 'app setup phase' (above the store) and and 'operation phase' showing what happens when the user requests an action. It would be useful to show more clearly the sequence of events that occur when the user requests repos in the example app.
- It would be helpful to list on the diagram the source files which correspond to the blocks on the diagram.

2. Under **Workflow** you provide a good walkthrough of the app in action. If you could correlate the discussion here with blocks and notes on the diagram(s) that would provide additional value.

**Comment by KarandikarMihir** [comment]
Thanks! 😄 I'll sit on this for a while. If there are any other changes, I can make those in a single batch.

**Comment by web2style** [comment]
In `Teck Stack -> Core` section I'd add that Redux Saga, Reselect, ImmutableJS and Styled Components are optional and there are instructions on how to remove them in the docs.

File: docs/general/introduction.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+## Project Structure
+Let's start with understanding why this boilerplate has chosen this particular structure. It has been an [evolving discussion](https://github.com/mxstbr/react-boilerplate/issues/27). Your source code will reside inside `app` folder. Configuration and reside inside `internals` folder. `server` contains development and production server configuration files.
+
+### `app/`
+We use [container/component architecture](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0#.4rmjqneiw). `containers/` contains React components which are connected to the redux store. `components/` contains dumb React components which depend on containers for data. Generally, you'd want to treat one webpage (such as Login, Home etc.) as one container and a small part (such as Login form, Navigation bar) of that page as a component. But there are no rigid rules. You can bend this structure to your needs.
+
+### `internals/`
+You can call this area as the engine of your `app/`. Your source code cannot be executed as-is in the web broweser. It needs to pass through webpack to get converted into a form that web browsers understand.
```
Comment (lamoglia) [inline docs/general/introduction.md]: Typo here on "broweser"

File: docs/general/introduction.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+### React Router:
+`<Router />` has the crucial information of your routes. Check out [`routes.js`](https://github.com/mxstbr/react-boilerplate/blob/master/app/routes.js) to see how route `path`s are mapped with application containers.
+
+- Path `"/"` corresponds to container `<Home />`
+- Path `"/features"` corresponds to container `<FeaturePage />`
+- Path `"*"` i.e. all other paths correspond to `<NotFoundPage />`
+
+These containers along with their corresponding reducer and sagas are loaded asynchronously with the help of `System.import()`. Whenever webpack encouters `System.import()` in the code, it creates a separate chunk for those imports. That means for every route, there will be a separate chunk. And by corollary, only those javascript chunks will be downloaded by the browser which are required for the current route. So when you navigate to `"/"`, only chunks related to `<Home />` will be downloaded and subsequently executed. This makes your application incredibly lightweight and lightening fast.
```
Comment (lamoglia) [inline docs/general/introduction.md]: Typo on "lightening"? Should it be "lightning"?

File: docs/general/introduction.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+- **Memoization:** Selectors will not compute a new result unless one of its arguments change. That means, if you are repeating the same search key, reselect will not filter the array again and again. It will just return the previous result. Reselect compares old and new arguments and decides whether to compute a new result.
+- **Composability:** You can combine multiple selectors. For eg. one selector can filter usernames according to search key and another selector can filter the already filtered array according to gender. One more selector can further filter according to age. You can combine three selectors by using `createSelector()`
+
+### Redux Saga:
+If your application is going to interact with some back-end application for data, we recommend using redux saga for side effect management. Too much of jargon? Let's simplify.
+
+Imagine that your application is fetching data in json format from a back-end. For a every API call, ideally you should define at least three kinds of [action creators](http://redux.js.org/docs/basics/Actions.html):
```
Comment (lamoglia) [inline docs/general/introduction.md]: Looks like the 'a' is not needed: "For a every API call"

File: docs/general/introduction.md (hunk 1 above)
Comment (KarandikarMihir) [inline docs/general/introduction.md]: Thank you for the review! Looks like @gihrig has taken care of this just now.

File: docs/general/introduction.md (hunk 2 above)
Comment (KarandikarMihir) [inline docs/general/introduction.md]: Done

File: docs/general/introduction.md (hunk 3 above)
Comment (KarandikarMihir) [inline docs/general/introduction.md]: Done

**Comment by KarandikarMihir** [comment]
@gihrig Very nice of you to make those changes! :+1: :smile: Thank you.

That leaves me with this-
- [ ] Split diagram into focused diagrams.
- [ ] Add sequence numbers in the diagrams.
- [ ] Relate `app/index.html` with `build/index.html`.
- [ ] Correlate building blocks and notes on diagrams.

**Comment by gihrig** [comment]
@KarandikarMihir @mxstbr 

>That leaves me with this-
 Split diagram into focused diagrams.
 Add sequence numbers in the diagrams.
 Relate app/index.html with build/index.html.
 Correlate building blocks and notes on diagrams.

IMO, Those items would be helpful, but their absence should not hold up the release of 3.4.

**Comment by gihrig** [comment]
@KarandikarMihir One more thing. Please host image file(s) on this repo.

**Comment by KarandikarMihir** [comment]
@gihrig OK. I'll add that image alongside `progressbar.md`. Other changes are not too small but not very big too. It'll take some time for me to write, for others to review, make iterations and so on. They can be included in .1 maybe?

**Comment by KarandikarMihir** [comment]
@gihrig I have a query. To host that image, we'll need a sort of permalink. Hosting it in my fork will not be safe, because the link will break if I ever delete the fork. Do we have to add image directly to master?

**Comment by gihrig** [comment]
@KarandikarMihir 

>To host that image, we'll need a sort of permalink.

Those image files need to be included in the same dir as your document and loaded with a relative link, as was done for WebStorm config [here](https://github.com/mxstbr/react-boilerplate/blob/master/docs/general/faq.md#using-this-boilerplate-with-webstorm). 

e.g. Line 193 `How to debug using WebStorm`

**Comment by KarandikarMihir** [comment]
@gihrig OK.. I'll try my best!

**Comment by KarandikarMihir** [comment]
@gihrig This fixes it. The reason I asked about permalink is because we have hosted the banner image that we see in `README.md` in a separate repo.. I thought we were going to do the same here :sweat_smile: 

You can merge this now. I don't think I can make those changes in time soon. This is good to go.

**Comment by gihrig** [comment]
@KarandikarMihir Well, that's (almost) perfect! 😄 

Just one _tiny_ change: Please rename `Workflow.png` to `workflow.png` (and also the corresponding  `<img src="Workflow.png"...`.

Then I'll go ahead and merge.

**Comment by KarandikarMihir** [comment]
@gihrig There you go! :)

**Comment by gihrig** [comment]
My feeling is that this is totally ready to go, no comments for two+ days. I know people are "away" for holidays, but since this is documentation, not affecting code, I'm going to go ahead and merge.

If anyone has further additions or corrections, a new PR is welcome.

**Pull request merged**